Add support for getting annotation default values

jameskleeh opened the pull request.

jameskleeh commented:
This completes part of https://github.com/google/ksp/issues/674 and https://github.com/google/ksp/issues/542

jameskleeh commented on kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSAnnotationImpl.kt:
```diff
@@ -27,6 +27,7 @@ class KSAnnotationImpl(private val annotationApplication: KtAnnotationApplicatio
     override val arguments: List<KSValueArgument> by lazy {
         annotationApplication.arguments.map { KSValueArgumentImpl(it) }
     }
+    override val defaultArguments: List<KSValueArgument> = emptyList()
```
Similar question as the other PR. I don't know if this is correct because I don't have context on this module

neetopia replied:
as mentioned in previous PR, you can put a TODO() here.

jameskleeh replied:
Done

jameskleeh commented:
@neetopia Any feedback so I can move this forward?

neetopia commented on compiler-plugin/testData/api/annotationWithDefaultValues.kt:
```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2020 Google LLC
```
Can you change this to 2022? Same as other occurrences.

jameskleeh replied:
Done

jameskleeh commented:
@neetopia Anything else I can do to move this forward?

neetopia commented on compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/binary/KSAnnotationDescriptorImpl.kt:
```diff
@@ -155,6 +159,13 @@ fun AnnotationDescriptor.createKSValueArguments(ownerAnnotation: KSAnnotation):
     return presentValueArguments.plus(argumentsFromDefault)
 }
 
+fun AnnotationDescriptor.getDefaultArguments(ownerAnnotation: KSAnnotation): List<KSValueArgument> {
```
minor: can you mark this `internal`?

neetopia commented on compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/java/KSAnnotationJavaImpl.kt:
```diff
@@ -97,6 +97,16 @@ class KSAnnotationJavaImpl private constructor(val psi: PsiAnnotation) : KSAnnot
         presentValueArguments.plus(argumentsFromDefault)
     }
 
+    override val defaultArguments: List<KSValueArgument> by lazy {
```
This shares quite a lot same logic as `fun AnnotationDescriptor.getDefaultArguments`, can you make this share the logic? A way to do this is to make `getDefaultArguments()` an extension function of KotlinType.

neetopia reviewed (commented):
overall LGTM, please also wait till I fix the API verification on main to rerun CI.

jameskleeh commented:
@neetopia This branch got all out of whack so I re-applied the changes on a fresh branch off of main and reset this one from the new one. I addressed the feedback, however the functions cannot be made internal because `AnnotationDescriptor.getDefaultArguments` is used from `com.google.devtools.ksp.symbol.impl.kotlin.KSAnnotationImpl` and the new function on `KotlinType` is used from `KSAnnotationJavaImpl`

neetopia commented:
> @neetopia This branch got all out of whack so I re-applied the changes on a fresh branch off of main and reset this one from the new one. I addressed the feedback, however the functions cannot be made internal because `AnnotationDescriptor.getDefaultArguments` is used from `com.google.devtools.ksp.symbol.impl.kotlin.KSAnnotationImpl` and the new function on `KotlinType` is used from `KSAnnotationJavaImpl`

that's not true though, `internal` in kotlin restricts access to same module, rather than the package private in Java, since they are in same module `compiler-plugin`, they are fine.

Can you also run `./gradlew :api:updateApi` to get metalava api signature test pass?

jameskleeh commented:
@neetopia Yeah that was my understanding as well. It appears IntelliJ was lying to me. I changed them to be internal and did the updateAPI call

neetopia commented:
Thanks for PR!
